packages/browser/src/iframe.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { ISessionInfo } from "@inrupt/solid-client-authn-core";
+
+/**
+ * Redirects the browser to a provided IRI, but does such redirection in a child
+ * iframe. This is used to have a front-channel interaction with the Solid Identity
+ * Provider without having the user involved, and without refreshing the main window.
+ *
+ * @param redirectUrl The IRI to which the iframe should be redirected.
+ */
+export function redirectInIframe(redirectUrl: string): void {
+  const iframe = window.document.createElement("iframe");
+  iframe.setAttribute("id", "token-renewal");
+  iframe.setAttribute("name", "token-renewal");
+  iframe.setAttribute("hidden", "true");
+  iframe.setAttribute("sandbox", "allow-scripts allow-same-origin");
+  window.document.body.appendChild(iframe);
+  iframe.src = redirectUrl;
+}
+
+/**
+ * This function sets up an event listener that will receive iframe messages.
+ * It is only listening to messages coming from iframes that could have been
+ * opened by the library, and expects to be posted the IRI the iframe has been
+ * redirected to by the Solid Identity Provider. This way, the top window can
+ * perform the backchannel exchange to the token endpoint without performing
+ * the front-channel redirection.
+ *
+ * @param handleIframeRedirect Redirect URL sent by the iframe
+ */
+export function setupIframeListener(
+  handleIframeRedirect: (
+    redirectUrl: string
+  ) => Promise<ISessionInfo | undefined>
+): void {
+  window.addEventListener("message", async (evt: MessageEvent) => {
+    const frameElement = document
+      .getElementsByTagName("iframe")
+      .namedItem("token-renewal");
```

Comment:
Unfortunately, I think this code needs to run before `redirectInIframe`: once `redirectinIframe` is called, the created DOM element will asynchronously message the main window, so if the redirection happens before setting up the listener, I'm afraid it could create race conditionswhere the auth code is sent before the listener is set up.

However, this only prevents from creating the DOM element in `redirectInIframe`, to return it from there, and somehow pass it here as an argument. I'll see if it is possible to create the DOM element outside the closure of both functions, and use side effects to do as you suggest.

Reply:
OK perfect!

Reply:
indeed, 07b65775863e9a259e6b216628bc6e3dc8f95e90 removes the iframe id or name, and only uses the local reference, which is cleaner I think.

Reply:
Ah I see, I thought it would be passed in as an argument to the function, but this probably works well. Only thing I'm wondering is that maybe this will break in server-side rendering use cases, since `window` isn't defined yet when importing.

One other approach could be to have a `getIframe` closure that only instantiates it when it's called, which would also get rid of a side effect:

```
let iframe: HTMLIFrameElement;
function getIframe(): HTMLIFrameElement {
  if (typeof iframe === "undefined") {
    iframe = window.document.createElement("iframe");
    iframe.setAttribute("hidden", "true");
    iframe.setAttribute("sandbox", "allow-scripts allow-same-origin");
  }
  return iframe;
}
```

Reply:
I didn't consider server-side rendering, so initializing the `iframe` in a separate closure is most definitively a good idea. I was thinking about introducing this design in the future anyway, because it's probable that the iframe will need to have different attributes depending on the browser (e.g. Mozilla introduced a special attribute to allow iframes to request access to cookies)